File: code/modM_MSepMarg.R
# Marginalized BYM2 model for M_M (MICS-only) data
# Uses single n-dimensional Epsilon_bym2 with eigendecomposition
# Eliminates the (w,u) funnel from the separable parameterization

fitMMMarg = function(datDHS=ed, datMICS=edMICS, inputsMDM=NULL, 
                     intPtsMICS=NULL, intPtsDHS=NULL, 
                     KMICS=100,
                     KDHSurb = 11,
                     JInnerUrban = 3,
                     KDHSrur = 16,
                     JInnerRural = 3,
                     JOuterRural = 1, admMICS=admFinal, adm2DHS=adm2Full, 
                     alpha_pri = c(0, 100^2), 
                     beta_pri = c(0, sqrt(1000)), 
                     pc.bym2Phi=list(u=0.5, alpha=1/3), 
                     pc.bym2Prec=list(u=1, alpha=0.5), 
                     pc.expPrec=list(u=1, alpha=0.5), 
                     covariates=NULL,
                     fixedHyperparams=NULL,
                     maxit=1000, adm2AsCovariate=FALSE, 
                     tolSeq = c(1e-06), getSDs=TRUE, doMCMC=FALSE) {
  
  # make sure Stratum variable exists in MICS data
  if(!("Stratum" %in% names(datMICS))) {
    datMICS$Stratum = adm2ToStratumMICS(datMICS$subarea)
  }
  
  # make sure all the names in our datasets we expect to be there are
  nameTab = rbind(c("N", "ns"), 
                  c("N", "n"), 
                  c("Z", "ys"), 
                  c("Z", "y"))
  for(i in 1:nrow(nameTab)) {
    theseNames = nameTab[i,]
    fromN = theseNames[1]
    toN = theseNames[2]
    if(!(toN %in% names(datMICS))) {
      datMICS[[toN]] = datMICS[[fromN]]
    }
    if(!(toN %in% names(datDHS))) {
      datDHS[[toN]] = datDHS[[fromN]]
    }
  }
  
  # first generate all necessary inputs if need be
  if(is.null(inputsMDM)) {
    print("Making M_DM inputs...")
    
    inputsMDM = makeInputsMDM(datDHS, datMICS, 
                              intPtsMICS=intPtsMICS, intPtsDHS=intPtsDHS, 
                              KMICS=KMICS,
                              KDHSurb = KDHSurb,
                              JInnerUrban = JInnerUrban,
                              KDHSrur = KDHSrur,
                              JInnerRural = JInnerRural,
                              JOuterRural = JOuterRural, 
                              admMICS=admMICS, adm2DHS=adm2DHS, 
                              adm2AsCovariate=adm2AsCovariate)
    
  }
  
  thisEnv = environment()
  list2env(inputsMDM, envir=thisEnv)
  
  datMICS = sortByCol(datMICS, "Stratum", admMICS$NAME_FINAL)
  
  # set priors ----
  
  out = load("savedOutput/global/admFinalMat.RData")
  bym2ArgsTMB = prepareBYM2argumentsForTMB(admFinalMat, u=pc.bym2Phi$u, alpha=pc.bym2Phi$alpha, 
                                           constr=TRUE, scale.model=TRUE, matrixType="TsparseMatrix")
  lambdaTau = getLambdaPCprec(u=pc.bym2Prec$u, alpha=pc.bym2Prec$alpha)
  lambdaTauEps = getLambdaPCprec(u=pc.expPrec$u, alpha=pc.expPrec$alpha)
  
  nAreas = ncol(bym2ArgsTMB$Q)
  
  # Subset covariates if requested ----
  allCovNames = colnames(intPtsMICS$XUrb)
  if(!is.null(covariates)) {
    keepIdx = which(allCovNames %in% covariates)
    cat("  Subsetting covariates to:", allCovNames[keepIdx], "\n")
    intPtsMICS$XUrb = intPtsMICS$XUrb[, keepIdx, drop=FALSE]
    intPtsMICS$XRur = intPtsMICS$XRur[, keepIdx, drop=FALSE]
  } else {
    cat("  Using all covariates:", allCovNames, "\n")
  }
  
  # Specify inputs for TMB ----
  
  ## specify random effects (alpha is explicit, not absorbed)
  rand_effs <- c('alpha', 'beta', 'Epsilon_bym2', 
                 'nuggetUrbMICS', 'nuggetRurMICS')
  
  # collect input data: use V_bym2 (eigenvectors) instead of Q_bym2
  data_full = list(
    y_iUrbanMICS=ysUrbMICS,
    y_iRuralMICS=ysRurMICS,
    n_iUrbanMICS=nsUrbMICS,
    n_iRuralMICS=nsRurMICS,
    areaidxlocUrbanMICS=areaidxlocUrbanMICS,
    areaidxlocRuralMICS=areaidxlocRuralMICS,
    X_betaUrbanMICS=intPtsMICS$XUrb,
    X_betaRuralMICS=intPtsMICS$XRur,
    wUrbanMICS=intPtsMICS$wUrban,
    wRuralMICS=intPtsMICS$wRural,
    V_bym2=bym2ArgsTMB$V,       # eigenvectors of Q (NOT the sparse Q itself)
    alpha_pri=alpha_pri,
    beta_pri=beta_pri,
    tr=bym2ArgsTMB$tr,
    gammaTildesm1=bym2ArgsTMB$gammaTildesm1,
    lambdaPhi=bym2ArgsTMB$lambda,
    lambdaTau=lambdaTau,
    lambdaTauEps=lambdaTauEps, 
    options=0
  )
  
  
  anyna = function(x) {any(is.na(x))}
  myDim = function(x) {
    dims = dim(x)
    if(is.null(dims)) {
      length(x)
    }
    else {
      dims
    }
  }
  
  # initial parameters
  initUrbP = sum(c(data_full$y_iUrbanMICS))/sum(c(data_full$n_iUrbanMICS))
  initRurP = sum(c(data_full$y_iRuralMICS))/sum(c(data_full$n_iRuralMICS))
  initAlpha = logit(initRurP)
  initBeta1 = logit(initUrbP) - initAlpha
  
  tmb_params <- list(log_tau = 0,
                     logit_phi = 0,
                     log_tauEps = 0,
                     alpha = initAlpha,
                     beta = c(initBeta1, rep(0, ncol(data_full$X_betaUrbanMICS)-1)), 
                     Epsilon_bym2 = rep(0, nAreas),
                     nuggetUrbMICS = rep(0, length(data_full$y_iUrbanMICS)), 
                     nuggetRurMICS = rep(0, length(data_full$y_iRuralMICS))
  )
  
  # Handle fixed hyperparameters via map ----
  mapList <- list()
  if(!is.null(fixedHyperparams)) {
    for(pName in names(fixedHyperparams)) {
      tmb_params[[pName]] <- fixedHyperparams[[pName]]
      mapList[[pName]] <- factor(NA)
      cat("  Fixing", pName, "=", fixedHyperparams[[pName]], "\n")
    }
  }
  
  # make TMB fun and grad ----
  
  # first make sure no TMB dynlib is loaded
  unloadDynlibs()
  
  # compile dynlib if need be
  if(!any(file.exists(paste0("code/modM_MSepMarg", c(".o", ".so", ".dll"))))) {
    print("compiling code/modM_MSepMarg.cpp...")
    compile( "code/modM_MSepMarg.cpp", 
             framework="TMBad", safebounds=FALSE)
  }
  
  # load dynlib, make tmb obj
  dyn.load( dynlib("code/modM_MSepMarg"))
  TMB::config(tmbad.sparse_hessian_compress = 1)
  obj <- MakeADFun(data=data_full,
                   parameters=tmb_params,
                   random=rand_effs,
                   map=mapList,
                   hessian=TRUE,
                   DLL='modM_MSepMarg')
  
  lower = rep(-10, length(obj[['par']]))
  upper = rep( 10, length(obj[['par']]))
  
  # make wrapper functions that print out parameters and function values
  funWrapper = function(par, badParVal=1e10) {
    if(any(par < lower) || any(par > upper)) {
      return(badParVal)
    }
    print(par)
    objVal = testObj[['fn']](par)
    parNames = names(par)
    parVals = par
    parStrs = sapply(1:length(par), function(ind) {paste(parNames[ind], ": ", parVals[ind], sep="")})
    parStr = paste(parStrs, collapse=", ")
    print(paste0("objective: ", objVal, " for parameters, ", parStr))
    objVal
  }
  
  grWrapper = function(par, badParVal=1e10) {
    if(any(par < lower) || any(par > upper)) {
      return(rep(badParVal, length(par)))
    }
    print(par)
    grVal = testObj[['gr']](par)
    parNames = names(par)
    parVals = par
    parStrs = sapply(1:length(par), function(ind) {paste(parNames[ind], ": ", parVals[ind], sep="")})
    parStr = paste(parStrs, collapse=", ")
    grStr = paste(grVal, collapse=", ")
    print(paste0("gradient: ", grStr, " for parameters, ", parStr))
    grVal
  }
  
  # * Run TMB ----
  
  if(!doMCMC) {
    {
      testObj = obj
      optPar = testObj$par
      startTime = proc.time()[3]
      for(thisTol in tolSeq) {
        testObj = obj
        testObj$env$inner.control = list(maxit=maxit, tol10=thisTol)
        testObj$env$tracepar = TRUE
        print(paste0("optimizing for tol = ", thisTol, "."))
        opt1 <- optim(par=optPar, fn=funWrapper, gr=grWrapper,
                      method = c("BFGS"), hessian = FALSE, control=list(reltol=thisTol))
        optPar = opt1$par
        if(!is.null(opt1$message)) {
          print(paste0("error for tol = ", thisTol, ". Message:"))
          print(opt1$message)
          next
        }
        else {
          print(paste0("completed optimization for tol = ", thisTol, ""))
          
          if(getSDs) {
            ## Get standard errors
            print("getting standard errors...")
            sdTime = system.time(
              SD0 <- TMB::sdreport(testObj, getJointPrecision=TRUE,
                                   bias.correct = TRUE,
                                   bias.correct.control = list(sd = TRUE))
            )[3]
            
            print(sdTime/60)
            
            if(SD0$pdHess) {
              print("Optimization and PD hess calculation done!")
              break
            }
            else {
              print("Hessan not PD. Rerunning optimization with stricter tol...")
            }
          } else {
            ## Get report without computing standard errors
            print("getting report (no standard errors)...")
            sdTime = system.time(
              SD0 <- testObj$report()
            )[3]
            print(paste0("report time: ", sdTime/60, " minutes"))
          }
        }
      }
      endTime = proc.time()[3]
      sdTime/60
      totalTime = endTime - startTime
      print(paste0("optimization took ", totalTime/60, " minutes"))
    }
  } else {
    {
      testObj = obj
      optPar = testObj$par
      startTime = proc.time()[3]
      
      # Run MCMC with adequate iterations for proper inference
      fit <- tmbstan(obj=obj, silent=FALSE, laplace=FALSE, 
                     iter=4000, warmup=2000, chains=1)
      
      endTime = proc.time()[3]
      totalTime = endTime - startTime
      print(paste0("MCMC took ", totalTime/60, " minutes"))
    }
  }
  
  
  list(TMBobj=obj, TMBsd=SD0, totalTime=totalTime, sdTime=sdTime)
}

File: code/modM_MSepMarg.cpp
// Marginalized BYM2 model for M_M (MICS-only) data
// Uses single n-dimensional Epsilon_bym2 with eigendecomposition V_bym2
// Eliminates the (w,u) funnel from the separable parameterization
//
// include libraries
#include <TMB.hpp>
#include <Eigen/Sparse>
#include <vector>
using namespace density;
using Eigen::SparseMatrix;

// helper function for detecting NAs in the data supplied from R
template<class Type>
bool isNA(Type x){
  return R_IsNA(asDouble(x));
}

// PC prior on BYM2 tau (precision) parameter (on log scale)
template<class Type>
Type dPCPriTau(Type logTau, Type lambda)
{
  Type tau = exp(logTau);
  
  Type ldensity = log(lambda) - log(2.0) -Type(1.5)*logTau - lambda/sqrt(tau);
  
  // add in log Jacobian
  Type ljacobian = logTau;
  
  // get final log density
  Type lPriorTau = ldensity + ljacobian;
  
  return lPriorTau;
}

///////////////////////////
// the main function     //
// to calculate the jnll //
///////////////////////////
template<class Type>
Type objective_function<Type>::operator() ()
{
  
  // ~~~~~~~~~------------------------------------------------------~~
  // FIRST, we define params/values/data that will be passed in from R
  // ~~~~~~~~~~~------------------------------------------------------
  
  // Data (all except for X_ij is a vector of length num_i)
  DATA_VECTOR( y_iUrbanMICS );   // obs per binomial experiment at point i (clust)
  DATA_VECTOR( y_iRuralMICS );
  DATA_VECTOR( n_iUrbanMICS );   // Trials per cluster
  DATA_VECTOR( n_iRuralMICS );
  DATA_IVECTOR(areaidxlocUrbanMICS); // vector of areal indices
  DATA_IVECTOR(areaidxlocRuralMICS);
  DATA_MATRIX( X_betaUrbanMICS );  // design matrix
  DATA_MATRIX( X_betaRuralMICS );
  DATA_ARRAY( wUrbanMICS ); // nObsUrban x nIntegrationPointsUrban weight matrix
  DATA_ARRAY( wRuralMICS );
  
  DATA_MATRIX( V_bym2 ); // Q_bym2 = V Lambda V^T is the eigendecomposition of Q_bym2
  
  // prior parameters
  DATA_VECTOR( alpha_pri );
  DATA_VECTOR( beta_pri );
  
  // BYM2 and prior precomputed values
  DATA_SCALAR( tr );
  DATA_VECTOR( gammaTildesm1 );
  
  DATA_SCALAR( lambdaPhi );
  DATA_SCALAR( lambdaTau );
  DATA_SCALAR( lambdaTauEps );
  
  // Options
  DATA_SCALAR( options );
  // options == 1 : adreport spatial params
  
  // Fixed effects
  PARAMETER( log_tau );
  PARAMETER( logit_phi );
  PARAMETER( log_tauEps );
  
  // Intercept (explicit, not absorbed into spatial field)
  PARAMETER( alpha );
  PARAMETER_VECTOR( beta );
  
  // BYM2 effects: marginalized n-dimensional representation
  // Epsilon_bym2 ~ N(0, (1/tau) * [(1-phi)*I + phi*Q_besag^+])
  PARAMETER_VECTOR( Epsilon_bym2 );
  PARAMETER_VECTOR( nuggetUrbMICS );
  PARAMETER_VECTOR( nuggetRurMICS );
  
  // ~~~~~~~~~------------------------------------------------~~
  // SECOND, we define all other objects that we need internally
  // ~~~~~~~~~------------------------------------------------~~
  
  int num_iUrbanMICS = y_iUrbanMICS.size();
  int num_iRuralMICS = y_iRuralMICS.size();
  int nAreas = Epsilon_bym2.size();
  int n_integrationPointsUrbanMICS = wUrbanMICS.cols();
  int n_integrationPointsRuralMICS = wRuralMICS.cols();
  
  // Transform parameters
  Type sqrtTau = exp(Type(0.5) * log_tau);
  Type tau = exp(log_tau);
  Type phi = Type(1.0)/(Type(1.0) + exp(-logit_phi));
  Type sigmaEps = exp(Type(-0.5) * log_tauEps);
  
  // objective function -- joint negative log-likelihood/posterior
  Type jnll = 0.0;
  
  // ~~~~~~~~~------------------------------------------------~~-
  // THIRD, we calculate the contribution to the likelihood from:
  // 1) BYM2 GMRF density
  // 2) priors
  // 3) likelihood
  // ~~~~~~~~~------------------------------------------------~~-
  
  /////////
  // (1) //
  /////////
  
  // Marginalized BYM2 density using eigendecomposition
  // Epsilon ~ N(0, (1/tau) * [(1-phi)*I + phi*Q^+])
  // Precision: tau * [(1-phi)*I + phi*Q^+]^{-1}
  //          = tau * V * diag(1/(1 + phi*gammaTildesm1)) * V^T
  // Quadratic form: sum_i (sqrt(tau) * (Eps^T V)_i / sqrt(1 + phi*gammaTildesm1_i))^2
  
  matrix<Type> transformedEpsilon = Epsilon_bym2.matrix().transpose() * V_bym2;
  Type quad = 0.0;
  for(int i = 0; i < transformedEpsilon.size(); i++) {
    quad += pow(transformedEpsilon(0,i) * sqrtTau / sqrt(Type(1.0) + phi*gammaTildesm1(i)), 2.0);
  }
  
  // Log determinant of covariance: log|(1/tau)[(1-phi)I + phi*Q^+]|
  //   = sum(log(1 + phi*gammaTildesm1)) + n*(-log_tau)
  Type logDet = Type(0.0);
  for(int i = 0; i < gammaTildesm1.size(); i++) {
    logDet += log(Type(1.0) + phi*gammaTildesm1(i));
  }
  Type logDetTau = logDet + Type(nAreas) * (-log_tau);
  
  // BYM2 log density (dropping 2*pi constant)
  Type bym2LogLik = Type(-0.5) * logDetTau;
  bym2LogLik += Type(-0.5) * quad;
  jnll -= bym2LogLik;
  
  // add in iid nugget model
  Type nuggetLogLik = 0.0;
  for(int i = 0; i< nuggetUrbMICS.size(); i++) {
    nuggetLogLik += dnorm(nuggetUrbMICS(i), Type(0), sigmaEps, true);
  }
  for(int i = 0; i< nuggetRurMICS.size(); i++) {
    nuggetLogLik += dnorm(nuggetRurMICS(i), Type(0), sigmaEps, true);
  }
  jnll -= nuggetLogLik;
  
  /////////
  // (2) //
  /////////
  // Prior contributions to joint likelihood
  
  // PC prior for tau
  Type logPriTauEps = dPCPriTau(log_tauEps, lambdaTauEps);
  
  Type firstPt = log(lambdaTau) - log(2.0);
  Type secondPt = -Type(1.5)*log_tau;
  Type thirdPt = - lambdaTau/sqrt(tau);
  Type ldensityTau = firstPt + secondPt + thirdPt;
  Type ljacobianTau = log_tau;
  Type logPriTau = ldensityTau + ljacobianTau;
  
  // PC prior for phi
  Type n = gammaTildesm1.size();
  Type KLD = Type(0.5) * (phi * tr - phi * n - logDet);
  Type d = sqrt(Type(2.0) * KLD);
  Type lexpDensity = log(lambdaPhi) - lambdaPhi * d;
  
  Type sumVal = 0.0;
  for(int i = 0; i < n; i++) {
    sumVal += gammaTildesm1(i)/(1 + gammaTildesm1(i) * phi);
  }
  Type ljacobian = - log(d) - log(2.0) + log(tr - n - sumVal);
  ljacobian = ljacobian - logit_phi - Type(2.0) * log(Type(1.0) + exp(-logit_phi));
  Type logPriPhi = lexpDensity + ljacobian;
  
  jnll -= logPriTau;
  jnll -= logPriTauEps;
  jnll -= logPriPhi;
  
  // prior for intercept
  jnll -= dnorm(alpha, alpha_pri(0), alpha_pri(1), true);
  
  // prior for covariates
  for(int i = 0; i < beta.size(); i++) {
    jnll -= dnorm(beta(i), beta_pri(0), beta_pri(1), true);
  }
  
  /////////
  // (3) //
  /////////
  // jnll contribution from each datapoint i
  
  // Fixed effects at data points (includes alpha)
  vector<Type> fe_iUrbanMICS = X_betaUrbanMICS * beta;
  vector<Type> fe_iRuralMICS = X_betaRuralMICS * beta;
  
  vector<Type> latentFieldUrbMICS(num_iUrbanMICS * n_integrationPointsUrbanMICS);
  vector<Type> latentFieldRurMICS(num_iRuralMICS * n_integrationPointsRuralMICS);
  int thisIndex;
  Type thisWeight;
  Type thislik;
  array<Type> liksUrbMICS(num_iUrbanMICS, n_integrationPointsUrbanMICS);
  array<Type> liksRurMICS(num_iRuralMICS, n_integrationPointsRuralMICS);
  
  for (int obsI = 0; obsI < num_iUrbanMICS; obsI++) {
    thislik = 0.0;
    
    for (int intI = 0; intI < n_integrationPointsUrbanMICS; intI++) {
      thisIndex = num_iUrbanMICS * intI + obsI;
      
      // latent field: alpha + X*beta + Epsilon_bym2[area] + nugget
      latentFieldUrbMICS(thisIndex) = alpha + fe_iUrbanMICS(thisIndex) + Epsilon_bym2(areaidxlocUrbanMICS(obsI)) + nuggetUrbMICS(obsI);
      
      thisWeight = wUrbanMICS(obsI,intI);
      
      if(thisWeight > 0.0) {
        liksUrbMICS(obsI, intI) = dbinom_robust( y_iUrbanMICS(obsI), n_iUrbanMICS(obsI), latentFieldUrbMICS(thisIndex), false);
        thislik += thisWeight*liksUrbMICS(obsI, intI);
      }
      else {
        liksUrbMICS(obsI, intI) = thisWeight;
      }
      
    }
    
    if(thislik > 0.0) {
      jnll -= log(thislik);
    }
  }
  
  for (int obsI = 0; obsI < num_iRuralMICS; obsI++) {
    thislik = 0.0;
    
    for (int intI = 0; intI < n_integrationPointsRuralMICS; intI++) {
      thisIndex = num_iRuralMICS * intI + obsI;
      
      latentFieldRurMICS(thisIndex) = alpha + fe_iRuralMICS(thisIndex) + Epsilon_bym2(areaidxlocRuralMICS(obsI)) + nuggetRurMICS(obsI);
      
      thisWeight = wRuralMICS(obsI,intI);
      
      if(thisWeight > 0.0) {
        liksRurMICS(obsI, intI) = dbinom_robust( y_iRuralMICS(obsI), n_iRuralMICS(obsI), latentFieldRurMICS(thisIndex), false);
        thislik += thisWeight*liksRurMICS(obsI, intI);
      }
      else {
        liksRurMICS(obsI, intI) = thisWeight;
      }
      
    }
    
    if(thislik > 0.0) {
      jnll -= log(thislik);
    }
  }
  
  // ~~~~~~~~~~~
  // ADREPORT
  // ~~~~~~~~~~~
  if(options==1){
    ADREPORT(log_tau);
    ADREPORT(logit_phi);
  }
  
  REPORT(Epsilon_bym2);
  REPORT(log_tau);
  REPORT(logit_phi);
  REPORT(log_tauEps);
  REPORT(alpha);
  REPORT(beta);
  REPORT(quad);
  REPORT(logDetTau);
  REPORT(bym2LogLik);
  REPORT(nuggetLogLik);
  REPORT(logPriTau);
  REPORT(ldensityTau);
  REPORT(ljacobianTau);
  REPORT(firstPt);
  REPORT(secondPt);
  REPORT(thirdPt);
  REPORT(logPriTauEps);
  REPORT(KLD);
  REPORT(lexpDensity);
  REPORT(ljacobian);
  REPORT(d);
  REPORT(logPriPhi);
  REPORT(liksUrbMICS);
  REPORT(liksRurMICS);
  REPORT(latentFieldUrbMICS);
  REPORT(latentFieldRurMICS);
  REPORT(fe_iUrbanMICS);
  REPORT(fe_iRuralMICS);
  
  
  return jnll;
  
}
